### PR TITLE
Upgrade to pyqldb v3.1.0

### PR DESCRIPTION
**Pull request by simonz-bq**
- Upgrade qldb driver dependency to v3.1.0
- Added flag to enable query statistics

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: qldbshell/outcome.py [hunk 1]
```diff
@@ -1,3 +1,16 @@
...
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
```
Comment (byronlin13) [inline qldbshell/outcome.py:13]: Do we need this Outcome enum if we have command enum? Can we combine the two?

File: qldbshell/qldb_shell.py [hunk 2]
```diff
@@ -47,27 +48,28 @@ class QldbShell:
...
         if profile:
             print(profile)
         print()
 
-        self._driver = pooled_driver
+        self._driver = driver
         try:
-            session = self._driver.get_session()
-            tables_result = session.execute_lambda(lambda txn: txn.execute_statement(TABLE_QUERY))
+            tables_result = self._driver.execute_lambda(lambda txn: txn.execute_statement(TABLE_QUERY))
```
Comment (byronlin13) [inline qldbshell/qldb_shell.py:63]: Why not use `list_tables()`

File: qldbshell/qldb_shell.py [hunk 3]
```diff
@@ -170,14 +174,12 @@ def default(self, line):
         elif (self._is_interactive_transaction is False) and (self._strip_text(line) == "commit"):
             print("'commit' can only be used on an active transaction")
         else:
-            session = self._driver.get_session()
             try:
-                print_result(session.execute_lambda(
-                    lambda x: x.execute_statement(line)))
+                print_result(self._driver.execute_lambda(
```
Comment (byronlin13) [inline qldbshell/qldb_shell.py:178]: This can be on one line

File: qldbshell/outcome.py (hunk 1 above)
Comment (simonz-bq) [reply qldbshell/outcome.py]: I didn't feel strongly enough to do it. Combined now.

File: qldbshell/qldb_shell.py (hunk 2 above)
Comment (simonz-bq) [reply qldbshell/qldb_shell.py]: Done.

File: qldbshell/qldb_shell.py (hunk 3 above)
Comment (simonz-bq) [reply qldbshell/qldb_shell.py]: Done.

File: qldbshell/__main__.py [hunk 4]
```diff
@@ -59,6 +55,12 @@ def main():
         help="Name of a profile speficified in aws credentials setup whose credentials we should use",
```
Comment (byronlin13) [inline qldbshell/__main__.py:55]: specified

File: qldbshell/__main__.py [hunk 5]
```diff
@@ -68,22 +70,22 @@ def main():
     )
     args = parser.parse_args()
 
-  # Setup logging
+    # Setup logging
     if args.verbose:
         loglevel = logging.DEBUG
```
Comment (byronlin13) [inline qldbshell/__main__.py:75]: log_level

File: qldbshell/__main__.py [hunk 6]
```diff
@@ -59,6 +55,12 @@ def main():
         help="Name of a profile speficified in aws credentials setup whose credentials we should use",
         action="store",
     )
+    parser.add_argument(
+        "-q",
+        "--query-stats",
+        help="Enable display of query statistics",
+        action="store_true"
```
Comment (byronlin13) [inline qldbshell/__main__.py:62]: Just to confirm, we want by default, the query stats enabled?

File: qldbshell/qldb_shell.py [hunk 7]
```diff
@@ -47,27 +45,30 @@ class QldbShell:
 
     :type profile: str
 
-    :type pooled_driver: PooledQldbDriver
+    :type driver: QldbDriver
 
     """
```
Comment (byronlin13) [inline qldbshell/qldb_shell.py:50]: Add `show_stats` docstring.

File: qldbshell/qldb_shell.py [hunk 8]
```diff
@@ -47,27 +45,30 @@ class QldbShell:
...
 
-    def __init__(self, profile="default", pooled_driver=None):
+    def __init__(self, profile="default", driver=None, show_stats=False):
```
Comment (byronlin13) [inline qldbshell/qldb_shell.py:52]: Hmm this sets show_stats=false but the argument parser will pass in true?

File: qldbshell/qldb_shell.py [hunk 9]
```diff
@@ -114,6 +116,21 @@ def cmdloop(self, ledger):
                 if text:
                     self.onecmd(text)
             except KeyboardInterrupt:
+                if self._transaction_thread and self._transaction_thread.is_alive():
+                    self._statement_queue.put(CommandContainer(Command.ABORT))
+                    self._transaction_thread.join()
+                    try:
+                        container_command = self._result_queue.get().command
+                        print(container_command)
```
Comment (byronlin13) [inline qldbshell/qldb_shell.py:124]: Do we need to print?

File: qldbshell/qldb_shell.py [hunk 10]
```diff
@@ -151,7 +168,15 @@ def do_EOF(self, line):
 
     @zero_noun_command
     def quit_shell(self, line):
```
Comment (byronlin13) [inline qldbshell/qldb_shell.py:170]: line parameter is not used

File: requirements.txt [hunk 11]
```diff
@@ -1,5 +1,6 @@
+pyqldb~=3.1.0
 argparse~=1.4.0
-amazon.ion~=0.5.0
+amazon.ion~=0.7.0
 boto3~=1.9
```
Comment (byronlin13) [inline requirements.txt:4]: The driver has boto3 pegged at [1.16.56
](https://github.com/awslabs/amazon-qldb-driver-python/blob/master/requirements.txt#L2)
should we do the same?

File: setup.py [hunk 12]
```diff
@@ -15,7 +15,8 @@
 from qldbshell import version
 
 requires = ['boto3>=1.9.237',
```
Comment (byronlin13) [inline setup.py:17]: Same here

File: qldbshell/__main__.py (hunk 4 above)
Comment (simonz-bq) [reply qldbshell/__main__.py:55]: Fixed

File: qldbshell/__main__.py (hunk 5 above)
Comment (simonz-bq) [reply qldbshell/__main__.py:75]: Fixed

File: qldbshell/__main__.py (hunk 6 above)
Comment (simonz-bq) [reply qldbshell/__main__.py:62]: They're disabled by default, and only enabled if you add the flag

File: qldbshell/qldb_shell.py (hunk 8 above)
Comment (simonz-bq) [reply qldbshell/qldb_shell.py:52]: It's the default parameter. Not 100% necessary but there nonetheless.

File: qldbshell/qldb_shell.py (hunk 10 above)
Comment (simonz-bq) [reply qldbshell/qldb_shell.py:170]: I believe these are in-built function signatures of the shell framework.

File: requirements.txt (hunk 11 above)
Comment (simonz-bq) [reply requirements.txt:4]: I dont think this matters as this isn't a library, and is the endpoint application.

File: setup.py (hunk 12 above)
Comment (simonz-bq) [reply setup.py:17]: Ditto

File: qldbshell/__main__.py [hunk 13]
```diff
@@ -56,9 +52,15 @@ def main():
     parser.add_argument(
         "-p",
         "--profile",
-        help="Name of a profile speficified in aws credentials setup whose credentials we should use",
+        help="Name of a profile specified in aws credentials setup whose credentials we should use",
```
Comment (yoroto) [inline qldbshell/__main__.py:55]: omg, well spotted...

**Review by yoroto — APPROVED** [review]

**Pull request merged**